### ISSUE TITLE
Potential memory leak on gpgsqlbackend

### ISSUE THREAD

**Issue by commx** [PowerDNS/pdns#4456]
A while ago I've upgraded from pdns 3.4.x to 4.0.x and have experienced memory leak issues since then. The pdns_server process' memory usage increases over time and gets killed by the OS after the memory is consumed entirely. I've monitored this using internal monitoring tools and can verify that this happens on all pdns server that run version 4.0.x (4.0.1 as distributed with Ubuntu in my case) in my environment.

I've ran valgrind with full memory leak check for five minutes to see where the problems originates from. It looks like SQL statements are not being freed correctly. Below a few outputs from the valgrind session:

```
==11825== 145,600 bytes in 728 blocks are definitely lost in loss record 3,998 of 4,011
==11825==    at 0x4C2DB8F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==11825==    by 0x87B85DC: PQmakeEmptyPGresult (in /usr/lib/x86_64-linux-gnu/libpq.so.5.8)
==11825==    by 0x87C3850: ??? (in /usr/lib/x86_64-linux-gnu/libpq.so.5.8)
==11825==    by 0x87BA5D7: PQgetResult (in /usr/lib/x86_64-linux-gnu/libpq.so.5.8)
==11825==    by 0x87BA92A: ??? (in /usr/lib/x86_64-linux-gnu/libpq.so.5.8)
==11825==    by 0x85A021F: SPgSQLStatement::execute() (in /usr/lib/x86_64-linux-gnu/pdns/libgpgsqlbackend.so)
==11825==    by 0x196BB9: GSQLBackend::lookup(QType const&, DNSName const&, DNSPacket*, int) (in /usr/sbin/pdns_server)
==11825==    by 0x2F40AB: UeberBackend::lookup(QType const&, DNSName const&, DNSPacket*, int) (in /usr/sbin/pdns_server)
==11825==    by 0x292494: PacketHandler::questionOrRecurse(DNSPacket*, bool*) (in /usr/sbin/pdns_server)
==11825==    by 0x29410C: PacketHandler::question(DNSPacket*) (in /usr/sbin/pdns_server)
==11825==    by 0x1C249B: MultiThreadDistributor<DNSPacket, DNSPacket, PacketHandler>::makeThread(void*) (in /usr/sbin/pdns_server)
==11825==    by 0x644B6F9: start_thread (pthread_create.c:333)
==11825== 
==11825== 333,600 bytes in 1,668 blocks are definitely lost in loss record 4,009 of 4,011
==11825==    at 0x4C2DB8F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==11825==    by 0x87B85DC: PQmakeEmptyPGresult (in /usr/lib/x86_64-linux-gnu/libpq.so.5.8)
==11825==    by 0x87C3850: ??? (in /usr/lib/x86_64-linux-gnu/libpq.so.5.8)
==11825==    by 0x87BA5D7: PQgetResult (in /usr/lib/x86_64-linux-gnu/libpq.so.5.8)
==11825==    by 0x87BA92A: ??? (in /usr/lib/x86_64-linux-gnu/libpq.so.5.8)
==11825==    by 0x85A021F: SPgSQLStatement::execute() (in /usr/lib/x86_64-linux-gnu/pdns/libgpgsqlbackend.so)
==11825==    by 0x196BB9: GSQLBackend::lookup(QType const&, DNSName const&, DNSPacket*, int) (in /usr/sbin/pdns_server)
==11825==    by 0x1D9721: DNSBackend::getSOA(DNSName const&, SOAData&, DNSPacket*) (in /usr/sbin/pdns_server)
==11825==    by 0x1A6950: GSQLBackend::getUnfreshSlaveInfos(std::vector<DomainInfo, std::allocator<DomainInfo> >*) (in /usr/sbin/pdns_server)
==11825==    by 0x2F348A: UeberBackend::getUnfreshSlaveInfos(std::vector<DomainInfo, std::allocator<DomainInfo> >*) (in /usr/sbin/pdns_server)
==11825==    by 0x2CC461: CommunicatorClass::slaveRefresh(PacketHandler*) (in /usr/sbin/pdns_server)
==11825==    by 0x1C43EA: CommunicatorClass::mainloop() (in /usr/sbin/pdns_server)
```

This issue may also exist on other backends, but I haven't tested so far.

**Comment by ahupowerdns** [comment]
Thanks, oddly enough you were the second person in 24 hours to report this. PR #4459 is intended to resolve the problem, but I still need an expert to look at it.

**Comment by pieterlexis** [comment]
fixed in #4459
